Added new kb article treelist-enable-checkbox-selection-only-edit-mode (#2217)
telerik/blazor-docs: knowledge-base/treelist-enable-checkbox-selection-only-edit-mode.md

> ---

Reviewer: The customer user the word "disable". In our KB, we use the word "enable" more often than "disable". No need to change this, but just saying that we should try to preserve the customer context and keywords when they are technically correct.

> title: Enabling Checkbox Selection Only in Edit Mode for Blazor TreeList

Reviewer: ```suggestion
title: Enable TreeList Checkbox Selection Only in Edit Mode
```

> description: Learn how to show the Checkbox Column in the Telerik Blazor TreeList only during the edit mode and hide it in display mode.

Reviewer: ```suggestion
description: Learn how to show the Checkbox Column in the Telerik Blazor TreeList only in edit mode and hide it in display mode.
```

type: how-to
page_title: How to Enable Checkbox Selection Only in Edit Mode in Blazor TreeList

> slug: treelist-enable-checkbox-selection-only-edit-mode

Reviewer: ```suggestion
slug: treelist-kb-enable-checkbox-selection-only-edit-mode
```

tags: treelist, blazor, checkbox, selection, edit mode, visibility, checkboxcolumn
res_type: kb
ticketid: 1657252
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>

>             <td>TreeList for Blazor</td>

Reviewer: Also for Grid

>             <td>TreeList for Blazor</td>

Reviewer: ```suggestion
            <td>TreeList for Blazor, <br/ > Grid for Blazor</td>
```

        </tr>
    </tbody>
</table>

## Description

> I want to disable the Checkbox selection in display mode and allow the user to select items only when editing. Is it possible to hide the Checkbox Column in display mode and only show it in the edit mode?

Reviewer: Integrate this sentence into the bullet list.

This KB article also answers the following questions:
- How can I toggle the visibility of the Checkbox Column in the Blazor TreeList?
- Is it possible to enable Checkbox selection only during edit mode in TreeList?
- Can I hide the Checkbox Column in display mode for the TreeList?

## Solution

To enable the Checkbox selection only in edit mode in the TreeList component, follow these steps:

1. Enable the [`CheckBoxOnlySelection` parameter]({%slug treelist-columns-checkbox%}#parameters) of the `TreeListCheckboxColumn` to prevent users from selecting rows with a click outside of edit mode.
2. Bind the [`Visible` parameter]({%slug treelist-columns-checkbox%}#parameters) of the `TreeListCheckboxColumn` to a variable. This allows toggling its visibility based on whether the TreeList is in edit mode.
3. Handle the CUD (Create, Update, Delete) events to manage the visibility of the `TreeListCheckboxColumn`.
    - For [`Inline`]({%slug treelist-editing-inline%}) and [`Popup`]({%slug treelist-editing-popup%}) edit modes, this approach works directly. In the [`Incell` edit mode the events flow is a bit different]({%slug treelist-editing-incell%}), so you may need to adjust the logic.

> Below is an example implementation that toggles the visibility of the CheckboxColumn:

Reviewer 2: `CheckboxColumn` or Checkbox Column
```suggestion
Below is an example implementation that toggles the visibility of the Checkbox Column:
```

```CSHTML

> @using System.ComponentModel.DataAnnotations @* for the validation attributes *@

Reviewer: Validation is not relevant to the KB, so we can remove unnecessary stuff to make the example simpler and smaller.

        <TreeListColumn Field="Name" Expandable="true" Width="320px" />
        <TreeListColumn Field="Id" Editable="false" Width="120px" />
        <TreeListColumn Field="EmailAddress" Width="220px" />
        <TreeListColumn Field="HireDate" Width="220px" />

        <TreeListCommandColumn Width="280px">
            <TreeListCommandButton Command="Add" Icon="@SvgIcon.Plus">Add Child</TreeListCommandButton>
            <TreeListCommandButton Command="Edit" Icon="@SvgIcon.Pencil">Edit</TreeListCommandButton>
            <TreeListCommandButton Command="Delete" Icon="@SvgIcon.Trash">Delete</TreeListCommandButton>
            <TreeListCommandButton Command="Save" Icon="@SvgIcon.Save" ShowInEdit="true">Update</TreeListCommandButton>
            <TreeListCommandButton Command="Cancel" Icon="@SvgIcon.Cancel" ShowInEdit="true">Cancel</TreeListCommandButton>
        </TreeListCommandColumn>

    </TreeListColumns>
</TelerikTreeList>

@code {
    private List<Employee> Data { get; set; }
    private IEnumerable<Employee> SelectedItems { get; set; } = new List<Employee>();

    private bool SelectionEnabled { get; set; }    
    private bool ItemAlreadySelected { get; set; }    

>     // Sample CUD operations for the local data

Reviewer: Some comments of this sort are unnecessary, as the method names are self-evident. You can replace the necessary comments with `#region` if you like.

                                }); ;
                            LastId++;
                        }
                    }
                }

                _data[0].Name += " (non-editable, see OnEdit)";
            }

            return await Task.FromResult(_data);
        }

        public static async Task Update(Employee itemToUpdate)
        {
            UpdateItemRecursive(_data, itemToUpdate);
        }

        public static async Task Delete(Employee itemToDelete)
        {
            RemoveChildRecursive(_data, itemToDelete);
        }

        // sample helper methods for handling the view-model data hierarchy
        static void UpdateItemRecursive(List<Employee> items, Employee itemToUpdate)
        {
            for (int i = 0; i < items.Count; i++)
            {
                if (items[i].Id.Equals(itemToUpdate.Id))
                {
                    items[i] = itemToUpdate;
                    return;
                }

                if (items[i].DirectReports?.Count > 0)
                {
                    UpdateItemRecursive(items[i].DirectReports, itemToUpdate);
                }
            }
        }

>         static void RemoveChildRecursive(List<Employee> items, Employee item)

Reviewer: some missing accessors

        {
            for (int i = 0; i < items.Count(); i++)
            {
                if (item.Equals(items[i]))
                {
                    items.Remove(item);

                    return;
                }
                else if (items[i].DirectReports?.Count > 0)
                {
                    RemoveChildRecursive(items[i].DirectReports, item);

                    if (items[i].DirectReports.Count == 0)
                    {
                        items[i].HasChildren = false;
                    }
                }
            }
        }

        static void InsertItemRecursive(List<Employee> Data, Employee insertedItem, Employee parentItem)
        {
            insertedItem.Id = LastId++;
            if (parentItem != null)
            {
                parentItem.HasChildren = true;
                if (parentItem.DirectReports == null)
                {
                    parentItem.DirectReports = new List<Employee>();
                }

                parentItem.DirectReports.Insert(0, insertedItem);
            }
            else
            {
                Data.Insert(0, insertedItem);
            }
        }
    }
}
```

> Alternatively, if you prefer not to toggle the `CheckboxColumn` visibility but just disable it, you can apply conditional custom CSS based on the flag that you set in the `OnEdit`/`OnAdd` handlers.

Reviewer: Are we sure customers will be able to do this without an example from us?

Reviewer 3: Added one more example.
